File: node-tests/generated-config-policy.test.ts
import assert from "node:assert/strict";
import { join, resolve } from "node:path";
import test from "node:test";

import { DriftPolicy } from "../node-src/domain/drift-policy.js";
import {
  applyGeneratedConfigPolicies,
  applyGeneratedConfigPolicy,
  GeneratedConfigPolicyError,
} from "../node-src/domain/generated-config-policy.js";
import { loadPackRoot, type LoadedPackRoot } from "../node-src/metadata/loader.js";
import type { JsonObject } from "../node-src/metadata/validation.js";

const ROOT = process.cwd();
const PACKS_ROOT = resolve(
  process.env.INFRAWRIGHT_PACKS?.trim() || join(ROOT, "packs"),
);
const PACK_PROFILE = resolve(
  process.env.PACK_PROFILE?.trim() || join(ROOT, "packsets", "full.json"),
);

const SCHEMA: JsonObject = {
  block: {
    attributes: {
      description: { optional: true, type: "string" },
      filled: { optional: true, type: ["list", "string"] },
      name: { required: true, type: "string" },
      size_quota: { optional: true, type: "number" },
    },
    block_types: {
      rules: {
        nesting_mode: "list",
        block: {
          attributes: {
            label: { required: true, type: "string" },
            order: { optional: true, type: "number" },
          },
        },
      },
    },
  },
};

function root(override: JsonObject | null = null): LoadedPackRoot {
  return {
    loadResourceSchema: async () => SCHEMA,
    resources: new Map([["sample_resource", { override }]]),
  } as unknown as LoadedPackRoot;
}

function policy(resource: JsonObject): DriftPolicy {
  return new DriftPolicy({
    version: 1,
    resource_types: { sample_resource: resource },
  });
}

const ADDRESS = "sample_resource.iw_deadbeef";
const SIBLING_ADDRESS = "sibling_resource.iw_cafebabe";
const GENERATED = `resource "sample_resource" "iw_deadbeef" {
  description = "DROP"
  name        = "Example"

  rules {
    label = "one"
    order = 0
  }
  rules {
    label = "two"
    order = 2
  }
}
`;

test("generated config removes optional scalar and wildcard block leaves, then fills raw evidence", async () => {
  const selected = policy({
    projection_fill: [{
      path: "filled",
      source: "rawFilled",
      reason: "provider omitted it",
      approved_by: "unit",
    }],
    projection_omit: [{
      path: "description",
      reason: "provider default",
      approved_by: "unit",
    }],
    projection_omit_if: [{
      path: "rules[*].order",
      values: [0],
      reason: "sentinel",
      approved_by: "unit",
    }],
  });
  const result = await applyGeneratedConfigPolicy({
    addressToKey: new Map([[ADDRESS, "example"]]),
    generatedConfig: GENERATED,
    policy: selected,
    rawItems: new Map([["example", { rawFilled: ["one", "two"] }]]),
    resourceType: "sample_resource",
    root: root(),
  });
  assert.equal(result.edits, 3);
  assert.equal(result.text.includes("description"), false);
  assert.equal(result.text.includes("order = 0"), false);
  assert.equal(result.text.includes("order = 2"), true);
  assert.equal(result.text.includes('filled = [\n    "one",\n    "two",\n  ]'), true);
  assert.deepEqual(selected.staleEntries(), []);
});

test("pack drop_if_default removes provider-emitted sentinel without a drift policy", async () => {
  const generated = GENERATED.replace(
    '  description = "DROP"\n',
    "  size_quota = 0\n",
  );
  const result = await applyGeneratedConfigPolicy({
    addressToKey: new Map([[ADDRESS, "example"]]),
    generatedConfig: generated,
    policy: null,
    resourceType: "sample_resource",
    root: root({ drop_if_default: { size_quota: 0 } }),
  });
  assert.equal(result.edits, 1);
  assert.equal(result.text.includes("size_quota"), false);
  assert.equal(result.text.includes('name        = "Example"'), true);
});

test("pack drop_if_default preserves nonmatching provider values", async () => {
  const generated = GENERATED.replace(
    '  description = "DROP"\n',
    "  size_quota = 10\n",
  );
  const result = await applyGeneratedConfigPolicy({
    addressToKey: new Map([[ADDRESS, "example"]]),
    generatedConfig: generated,
    policy: null,
    resourceType: "sample_resource",
    root: root({ drop_if_default: { size_quota: 0 } }),
  });
  assert.deepEqual(result, { edits: 0, text: generated });
});

test("committed ZIA pack defaults remove only exact empty enums from generated config", async () => {
  const committed = await loadPackRoot({
    packsRoot: PACKS_ROOT,
    profilePath: PACK_PROFILE,
    catalogPath: PACK_PROFILE,
  });
  const cases = [{
    resourceType: "zia_dlp_dictionaries",
    edits: 2,
    generated: `resource "zia_dlp_dictionaries" "iw_deadbeef" {
  confidence_level_for_predefined_dict = ""
  confidence_threshold                 = ""
}
`,
  }, {
    resourceType: "zia_http_header_profile",
    edits: 2,
    generated: `resource "zia_http_header_profile" "iw_deadbeef" {
  name = "Header"

  http_header_profile_criteria {
    header     = "USERAGENT"
    operator   = ""
    user_agent = ""
  }
}
`,
  }, {
    resourceType: "zia_location_management",
    edits: 3,
    generated: `resource "zia_location_management" "iw_deadbeef" {
  display_time_unit           = ""
  name                        = "Location"
  sub_loc_scope               = ""
  surrogate_refresh_time_unit = ""
}
`,
  }, {
    resourceType: "zia_ssl_inspection_rules",
    edits: 1,
    generated: `resource "zia_ssl_inspection_rules" "iw_deadbeef" {
  name  = "SSL"
  order = 1

  action {
    type = "DO_NOT_DECRYPT"

    do_not_decrypt_sub_actions {
      bypass_other_policies = true
      min_tls_version       = ""
    }
  }
}
`,
  }] as const;

  for (const fixture of cases) {
    const address = `${fixture.resourceType}.iw_deadbeef`;
    const empty = await applyGeneratedConfigPolicy({
      addressToKey: new Map([[address, "example"]]),
      generatedConfig: fixture.generated,
      policy: null,
      resourceType: fixture.resourceType,
      root: committed,
    });
    assert.equal(empty.edits, fixture.edits, `${fixture.resourceType} empty edits`);
    const nonempty = await applyGeneratedConfigPolicy({
      addressToKey: new Map([[address, "example"]]),
      generatedConfig: fixture.generated.replaceAll('= ""', '= "SET"'),
      policy: null,
      resourceType: fixture.resourceType,
      root: committed,
    });
    assert.equal(nonempty.edits, 0, `${fixture.resourceType} nonempty edits`);
    const nullable = await applyGeneratedConfigPolicy({
      addressToKey: new Map([[address, "example"]]),
      generatedConfig: fixture.generated.replaceAll('= ""', "= null"),
      policy: null,
      resourceType: fixture.resourceType,
      root: committed,
    });
    assert.equal(nullable.edits, 0, `${fixture.resourceType} null edits`);
  }
});

test("pack defaults precede overlapping conditional drift omissions", async () => {
  const generated = GENERATED.replace(
    '  description = "DROP"\n',
    "  size_quota = 0\n",
  );
  const selected = policy({
    projection_omit_if: [{
      path: "size_quota",
      values: [0],
      reason: "overlap",
      approved_by: "unit",
    }],
  });
  const result = await applyGeneratedConfigPolicy({
    addressToKey: new Map([[ADDRESS, "example"]]),
    generatedConfig: generated,
    policy: selected,
    resourceType: "sample_resource",
    root: root({ drop_if_default: { size_quota: 0 } }),
  });
  assert.equal(result.edits, 1);
  assert.equal(result.text.includes("size_quota"), false);
  assert.equal(selected.staleEntries().length, 1);
});

test("batch generated-config policy edits known sibling resource blocks independently", async () => {
  const siblingGenerated = GENERATED.replaceAll("sample_resource", "sibling_resource")
    .replaceAll("iw_deadbeef", "iw_cafebabe")
    .replaceAll('description = "DROP"', 'description = "SIBLING"');
  const firstPolicy = policy({
    projection_omit: [{ path: "description", reason: "first", approved_by: "unit" }],
  });
  const siblingPolicy = new DriftPolicy({
    version: 1,
    resource_types: {
      sibling_resource: {
        projection_omit_if: [{
          path: "description",
          values: ["SIBLING"],
          reason: "second",
          approved_by: "unit",
        }],
      },
    },
  });
  const result = await applyGeneratedConfigPolicies({
    generatedConfig: `${GENERATED}\n${siblingGenerated}`,
    resources: [{
      addressToKey: new Map([[ADDRESS, "first"]]),
      policy: firstPolicy,
      resourceType: "sample_resource",
    }, {
      addressToKey: new Map([[SIBLING_ADDRESS, "second"]]),
      policy: siblingPolicy,
      resourceType: "sibling_resource",
    }],
    root: root(),
  });
  assert.equal(result.edits, 2);
  assert.equal(result.text.includes('description = "DROP"'), false);
  assert.equal(result.text.includes('description = "SIBLING"'), false);
  assert.deepEqual(firstPolicy.staleEntries(), []);
  assert.deepEqual(siblingPolicy.staleEntries(), []);
});

test("batch generated-config policy rejects unknown sibling blocks", async () => {
  const siblingGenerated = GENERATED.replaceAll("sample_resource", "sibling_resource")
    .replaceAll("iw_deadbeef", "iw_cafebabe");
  await assert.rejects(
    () => applyGeneratedConfigPolicies({
      generatedConfig: `${GENERATED}\n${siblingGenerated}\nresource "unknown_resource" "extra" {\n  name = "extra"\n}\n`,
      resources: [{
        addressToKey: new Map([[ADDRESS, "first"]]),
        policy: null,
        resourceType: "sample_resource",
      }, {
        addressToKey: new Map([[SIBLING_ADDRESS, "second"]]),
        policy: null,
        resourceType: "sibling_resource",
      }],
      root: root(),
    }),
    /unexpected resource block unknown_resource\.extra/u,
  );
});

test("no-policy single-resource wrapper preserves the schema-free fast path", async () => {
  const result = await applyGeneratedConfigPolicy({
    addressToKey: new Map([[ADDRESS, "example"]]),
    generatedConfig: GENERATED,
    policy: null,
    resourceType: "sample_resource",
    root: {
      loadResourceSchema: async () => {
        throw new Error("schema must not be loaded without policy entries");
      },
    } as unknown as LoadedPackRoot,
  });
  assert.deepEqual(result, { edits: 0, text: GENERATED });
});

test("exact-index generated-config omits are intentionally deferred to state projection", async () => {
  const selected = policy({
    projection_omit: [{
      path: "rules[0].order",
      reason: "exact plan index",
      approved_by: "unit",
    }],
  });
  const result = await applyGeneratedConfigPolicy({
    addressToKey: new Map([[ADDRESS, "example"]]),
    generatedConfig: GENERATED,
    policy: selected,
    resourceType: "sample_resource",
    root: root(),
  });
  assert.equal(result.edits, 0);
  assert.equal(result.text, GENERATED);
  assert.equal(selected.staleEntries().length, 1);
});

test("a value added by projection_fill is subsequently eligible for projection_omit_if", async () => {
  const selected = policy({
    projection_fill: [{
      path: "description",
      source: "rawDescription",
      reason: "provider omitted it",
      approved_by: "unit",
    }],
    projection_omit_if: [{
      path: "description",
      values: ["DROP"],
      reason: "sentinel",
      approved_by: "unit",
    }],
  });
  const result = await applyGeneratedConfigPolicy({
    addressToKey: new Map([[ADDRESS, "example"]]),
    generatedConfig: GENERATED.replace('  description = "DROP"\n', ""),
    policy: selected,
    rawItems: new Map([["example", { rawDescription: "DROP" }]]),
    resourceType: "sample_resource",
    root: root(),
  });
  assert.equal(result.edits, 2);
  assert.equal(result.text.includes("description"), false);
  assert.deepEqual(selected.staleEntries(), []);
});

test("required-path, missing-config, missing-raw, missing-address, duplicate, and unexpected blocks fail closed", async () => {
  await assert.rejects(
    () => applyGeneratedConfigPolicy({
      addressToKey: new Map([[ADDRESS, "example"]]),
      generatedConfig: GENERATED,
      policy: policy({ projection_omit: [{ path: "name", reason: "bad", approved_by: "unit" }] }),
      resourceType: "sample_resource",
      root: root(),
    }),
    /cannot projection_omit required path name/,
  );
  await assert.rejects(
    () => applyGeneratedConfigPolicy({
      addressToKey: new Map([[ADDRESS, "example"]]),
      generatedConfig: "",
      policy: policy({ projection_omit: [{ path: "description", reason: "test", approved_by: "unit" }] }),
      resourceType: "sample_resource",
      root: root(),
    }),
    /generated import config is missing/,
  );
  await assert.rejects(
    () => applyGeneratedConfigPolicy({
      addressToKey: new Map([[ADDRESS, "example"]]),
      generatedConfig: GENERATED,
      policy: policy({ projection_fill: [{ path: "filled", source: "rawFilled", reason: "test", approved_by: "unit" }] }),
      resourceType: "sample_resource",
      root: root(),
    }),
    /projection_fill requires raw_items/,
  );

  const active = policy({ projection_omit: [{ path: "description", reason: "test", approved_by: "unit" }] });
  await assert.rejects(
    () => applyGeneratedConfigPolicy({
      addressToKey: new Map([[ADDRESS, "example"], ["sample_resource.iw_missing", "missing"]]),
      generatedConfig: GENERATED,
      policy: active,
      resourceType: "sample_resource",
      root: root(),
    }),
    /missing resource block/,
  );
  await assert.rejects(
    () => applyGeneratedConfigPolicy({
      addressToKey: new Map([[ADDRESS, "example"]]),
      generatedConfig: `${GENERATED}${GENERATED}`,
      policy: active,
      resourceType: "sample_resource",
      root: root(),
    }),
    /duplicate resource block/,
  );
  await assert.rejects(
    () => applyGeneratedConfigPolicy({
      addressToKey: new Map([[ADDRESS, "example"]]),
      generatedConfig: GENERATED.replaceAll("iw_deadbeef", "unexpected"),
      policy: active,
      resourceType: "sample_resource",
      root: root(),
    }),
    /unexpected resource block/,
  );
});

test("unknown compound values are preserved rather than treated as omit-if scalar evidence", async () => {
  const generated = GENERATED.replace(
    'description = "DROP"',
    'description = upper("DROP")',
  );
  const selected = policy({
    projection_omit_if: [{
      path: "description",
      values: ["DROP"],
      reason: "only literal evidence is safe",
      approved_by: "unit",
    }],
  });
  const result = await applyGeneratedConfigPolicy({
    addressToKey: new Map([[ADDRESS, "example"]]),
    generatedConfig: generated,
    policy: selected,
    resourceType: "sample_resource",
    root: root(),
  });
  assert.equal(result.text, generated);
  assert.equal(result.edits, 0);
  assert.equal(selected.staleEntries().length, 1);
});

test("policy error type remains stable for callers", () => {
  const error = new GeneratedConfigPolicyError("example");
  assert.equal(error.name, "GeneratedConfigPolicyError");
});

File: go/internal/adopt/generated_config_policy_test.go
package adopt

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"

	"github.com/dvmrry/infrawright-dev/go/internal/metadata"
)

func generatedPolicyRoot(t *testing.T, override metadata.JsonObject) *metadata.LoadedPackRoot {
	t.Helper()
	root := testOracleRoot(t)
	manifest := root.Packs.Manifests[0]
	schemaDirectory := filepath.Join(manifest.Directory, "schemas", "provider")
	if err := os.MkdirAll(schemaDirectory, 0o700); err != nil {
		t.Fatalf("os.MkdirAll(schema): %v", err)
	}
	schema := map[string]any{"resource_schemas": map[string]any{
		testResourceType: map[string]any{"block": map[string]any{
			"attributes": map[string]any{
				"name":           map[string]any{"type": "string", "optional": true},
				"description":    map[string]any{"type": "string", "optional": true},
				"filled":         map[string]any{"type": "string", "optional": true},
				"required_value": map[string]any{"type": "string", "required": true},
			},
			"block_types": map[string]any{
				"optional_block": map[string]any{
					"nesting_mode": "list",
					"block": map[string]any{
						"attributes": map[string]any{
							"document": map[string]any{"type": "string", "optional": true},
							"name":     map[string]any{"type": "string", "required": true},
						},
						"block_types": map[string]any{
							"child": map[string]any{
								"nesting_mode": "list",
								"block": map[string]any{"attributes": map[string]any{
									"name": map[string]any{"type": "string", "required": true},
								}},
							},
						},
					},
				},
				"required_block": map[string]any{
					"min_items": json.Number("1"), "nesting_mode": "list",
					"block": map[string]any{"attributes": map[string]any{
						"name": map[string]any{"type": "string", "required": true},
					}},
				},
			},
		}},
	}}
	data, err := json.Marshal(schema)
	if err != nil {
		t.Fatalf("json.Marshal(schema): %v", err)
	}
	if err := os.WriteFile(filepath.Join(schemaDirectory, testProvider+".json"), data, 0o600); err != nil {
		t.Fatalf("os.WriteFile(schema): %v", err)
	}
	resource := root.Resources[testResourceType]
	resource.Override = override
	root.Resources[testResourceType] = resource
	return root
}

func testPolicy(t *testing.T, mode string, entries ...metadata.JsonObject) *metadata.DriftPolicy {
	t.Helper()
	raw := make([]any, len(entries))
	for index := range entries {
		raw[index] = entries[index]
	}
	policy, err := metadata.NewDriftPolicy(metadata.JsonObject{
		"version":        float64(1),
		"resource_types": metadata.JsonObject{testResourceType: metadata.JsonObject{mode: raw}},
	}, "generated policy test")
	if err != nil {
		t.Fatalf("metadata.NewDriftPolicy: %v", err)
	}
	return policy
}

func policyEntry(path string, extra metadata.JsonObject) metadata.JsonObject {
	entry := metadata.JsonObject{"path": path, "reason": "test", "approved_by": "unit"}
	for key, value := range extra {
		entry[key] = value
	}
	return entry
}

func TestGeneratedConfigProjectionOmitIsExactAndMarksEntry(t *testing.T) {
	policy := testPolicy(t, "projection_omit", policyEntry("description", nil))
	input := "resource \"test_item\" \"example\" {\n  name = \"fixture\"\n  description = \"drop me\"\n  required_value = \"keep\"\n}\n"
	result, err := ApplyGeneratedConfigPolicy(input, GeneratedConfigPolicyResource{
		AddressToKey: map[string]string{"test_item.example": "key"}, Policy: policy, ResourceType: testResourceType,
	}, generatedPolicyRoot(t, nil))
	if err != nil {
		t.Fatalf("ApplyGeneratedConfigPolicy: %v", err)
	}
	want := "resource \"test_item\" \"example\" {\n  name = \"fixture\"\n  required_value = \"keep\"\n}\n"
	if result.Edits != 1 || result.Text != want {
		t.Fatalf("result = %#v, want edits=1 text=%q", result, want)
	}
	if stale := policy.StaleEntries(metadata.StaleEntriesOptions{}); len(stale) != 0 {
		t.Fatalf("matched policy remained stale: %#v", stale)
	}
}

func TestGeneratedConfigProjectionOmitWholeOptionalBlock(t *testing.T) {
	policy := testPolicy(t, "projection_omit", policyEntry("optional_block", nil))
	input := "resource \"test_item\" \"example\" {\n" +
		"  optional_block {\n" +
		"    name = \"drop\"\n" +
		"    document = <<-EOT\n" +
		"      literal } and { text\n" +
		"    EOT\n" +
		"    child {\n" +
		"      name = \"nested\"\n" +
		"    }\n" +
		"  }\n" +
		"  required_value = \"keep\"\n" +
		"}\n"
	result, err := ApplyGeneratedConfigPolicy(input, GeneratedConfigPolicyResource{
		AddressToKey: map[string]string{"test_item.example": "key"}, Policy: policy, ResourceType: testResourceType,
	}, generatedPolicyRoot(t, nil))
	if err != nil {
		t.Fatalf("ApplyGeneratedConfigPolicy(optional block): %v", err)
	}
	want := "resource \"test_item\" \"example\" {\n  required_value = \"keep\"\n}\n"
	if result.Edits != 1 || result.Text != want {
		t.Errorf("ApplyGeneratedConfigPolicy(optional block) = %#v, want edits=1 text=%q", result, want)
	}
	if stale := policy.StaleEntries(metadata.StaleEntriesOptions{}); len(stale) != 0 {
		t.Errorf("ApplyGeneratedConfigPolicy(optional block) stale entries = %#v, want none", stale)
	}
}

func TestGeneratedConfigProjectionOmitRejectsRequiredBlock(t *testing.T) {
	policy := testPolicy(t, "projection_omit", policyEntry("required_block", nil))
	input := "resource \"test_item\" \"example\" {\n" +
		"  required_block {\n" +
		"    name = \"keep\"\n" +
		"  }\n" +
		"  required_value = \"keep\"\n" +
		"}\n"
	_, err := ApplyGeneratedConfigPolicy(input, GeneratedConfigPolicyResource{
		AddressToKey: map[string]string{"test_item.example": "key"}, Policy: policy, ResourceType: testResourceType,
	}, generatedPolicyRoot(t, nil))
	if err == nil || !strings.Contains(err.Error(), "non-optional block required_block") {
		t.Fatalf("ApplyGeneratedConfigPolicy(required block) error = %v, want required-block refusal", err)
	}
}

func TestCommittedZIA480GeneratedConfigOmitsEndpointBlocks(t *testing.T) {
	requireAdoptionPackSelection(t, metadata.PackSelection{Packs: []string{"zia"}, Shared: []string{"zscaler"}})
	root := installedAdoptionRoot(t)
	for _, resourceType := range []string{
		"zia_firewall_dns_rule",
		"zia_firewall_filtering_rule",
		"zia_firewall_ips_rule",
		"zia_ssl_inspection_rules",
	} {
		t.Run(resourceType, func(t *testing.T) {
			policy, err := LoadAdoptionPolicy(root, nil)
			if err != nil {
				t.Fatalf("LoadAdoptionPolicy(%s): %v", resourceType, err)
			}
			input := "resource \"" + resourceType + "\" \"example\" {\n" +
				"  name = \"Projected rule\"\n" +
				"  end_point_applications {\n" +
				"    zapp_id = \"zapp-1\"\n" +
				"  }\n" +
				"  end_point_application_groups {\n" +
				"    group_id = 42\n" +
				"  }\n" +
				"}\n"
			result, err := ApplyGeneratedConfigPolicy(input, GeneratedConfigPolicyResource{
				AddressToKey: map[string]string{resourceType + ".example": "key"},
				Policy:       policy, ResourceType: resourceType,
			}, &root)
			if err != nil {
				t.Fatalf("ApplyGeneratedConfigPolicy(%s): %v", resourceType, err)
			}
			if result.Edits != 2 || strings.Contains(result.Text, "end_point_application") {
				t.Errorf("ApplyGeneratedConfigPolicy(%s) = %#v, want both endpoint blocks omitted", resourceType, result)
			}
			stale := policy.StaleEntries(metadata.StaleEntriesOptions{ResourceTypes: map[string]struct{}{resourceType: {}}})
			if len(stale) != 0 {
				t.Errorf("ApplyGeneratedConfigPolicy(%s) stale entries = %#v, want none", resourceType, stale)
			}
		})
	}
}

func TestGeneratedConfigPackDefaultAndOmitIfOrder(t *testing.T) {
	override := metadata.JsonObject{"drop_if_default": metadata.JsonObject{"name": "pack-default"}}
	policy := testPolicy(t, "projection_omit_if", policyEntry("name", metadata.JsonObject{"values": []any{"pack-default"}}))
	input := "resource \"test_item\" \"example\" {\n  name = \"pack-default\"\n  required_value = \"keep\"\n}\n"
	result, err := ApplyGeneratedConfigPolicy(input, GeneratedConfigPolicyResource{
		AddressToKey: map[string]string{"test_item.example": "key"}, Policy: policy, ResourceType: testResourceType,
	}, generatedPolicyRoot(t, override))
	if err != nil {
		t.Fatalf("ApplyGeneratedConfigPolicy: %v", err)
	}
	if result.Edits != 1 || strings.Contains(result.Text, "name =") {
		t.Fatalf("result = %#v, want pack default removed first", result)
	}
	if stale := policy.StaleEntries(metadata.StaleEntriesOptions{}); len(stale) != 1 {
		t.Fatalf("overlapping policy stale count = %d, want 1", len(stale))
	}
}

func TestGeneratedConfigProjectionFillThenOmitIf(t *testing.T) {
	policy, err := metadata.NewDriftPolicy(metadata.JsonObject{
		"version": float64(1),
		"resource_types": metadata.JsonObject{testResourceType: metadata.JsonObject{
			"projection_fill":    []any{policyEntry("filled", metadata.JsonObject{"source": "rawFilled"})},
			"projection_omit_if": []any{policyEntry("filled", metadata.JsonObject{"values": []any{"sentinel"}})},
		}},
	}, "fill then omit test")
	if err != nil {
		t.Fatalf("metadata.NewDriftPolicy: %v", err)
	}
	input := "resource \"test_item\" \"example\" {\n  required_value = \"keep\"\n}\n"
	result, err := ApplyGeneratedConfigPolicy(input, GeneratedConfigPolicyResource{
		AddressToKey: map[string]string{"test_item.example": "key"},
		Policy:       policy,
		RawItems:     map[string]map[string]any{"key": {"rawFilled": "sentinel"}},
		ResourceType: testResourceType,
	}, generatedPolicyRoot(t, nil))
	if err != nil {
		t.Fatalf("ApplyGeneratedConfigPolicy fill then omit: %v", err)
	}
	if result.Edits != 2 || result.Text != input {
		t.Fatalf("fill-then-omit result = %#v, want edits=2 and original text", result)
	}
	if stale := policy.StaleEntries(metadata.StaleEntriesOptions{}); len(stale) != 0 {
		t.Fatalf("fill-then-omit stale entries = %#v, want none", stale)
	}
}

func TestGeneratedConfigRejectsRequiredOmitAndUnknownSibling(t *testing.T) {
	policy := testPolicy(t, "projection_omit", policyEntry("required_value", nil))
	_, err := ApplyGeneratedConfigPolicy("resource \"test_item\" \"example\" {\n}\n", GeneratedConfigPolicyResource{
		AddressToKey: map[string]string{"test_item.example": "key"}, Policy: policy, ResourceType: testResourceType,
	}, generatedPolicyRoot(t, nil))
	if err == nil || !strings.Contains(err.Error(), "required path") {
		t.Fatalf("required omit error = %v", err)
	}

	_, err = ApplyGeneratedConfigPolicies(ApplyGeneratedConfigPoliciesOptions{
		GeneratedConfig: "resource \"test_item\" \"example\" {\n}\n\nresource \"unknown\" \"sibling\" {\n}\n",
		Resources: []GeneratedConfigPolicyResource{
			{AddressToKey: map[string]string{"test_item.example": "key"}, ResourceType: testResourceType},
			{AddressToKey: map[string]string{"test_other.sibling": "key"}, ResourceType: "test_other"},
		},
		Root: generatedPolicyRoot(t, nil),
	})
	if err == nil || !strings.Contains(err.Error(), "unexpected resource block unknown.sibling") {
		t.Fatalf("unknown sibling error = %v", err)
	}
}

func TestGeneratedConfigCompoundExpressionIsPreserved(t *testing.T) {
	policy := testPolicy(t, "projection_omit_if", policyEntry("description", metadata.JsonObject{"values": []any{"x"}}))
	input := "resource \"test_item\" \"example\" {\n  description = coalesce(\n    \"x\",\n    \"y\",\n  )\n  required_value = \"keep\"\n}\n"
	result, err := ApplyGeneratedConfigPolicy(input, GeneratedConfigPolicyResource{
		AddressToKey: map[string]string{"test_item.example": "key"}, Policy: policy, ResourceType: testResourceType,
	}, generatedPolicyRoot(t, nil))
	if err != nil {
		t.Fatalf("ApplyGeneratedConfigPolicy compound: %v", err)
	}
	if result.Edits != 0 || result.Text != input {
		t.Fatalf("compound expression changed: %#v", result)
	}
}
